web-ui/src/pages/MainGame/MainGameScreen.tsx

```diff
@@ -5,33 +5,58 @@ import Container from "../../components/utils/Container";
 import CardBoard from "../../components/cards/CardBoard";
 import BiddingBoard from "../../components/bidding/BiddingBoard";
 import {InjectedProps} from "../../websocket/withWebsocketConnection";
-import {EventType, MoveType, PlayerBadeEvent, TopicTemplate,} from "../../websocket/events/types";
+import {MessageType, PlayerBadeEvent, TopicTemplate, MoveType, CardPlayedEvent} from "../../websocket/messages/types";
 import {
-  ContractBiddingMove,
-  ContractValue,
   GameRoundPhase,
   GameState,
   LegalBiddingMove,
   Position,
   SpecialBidding,
-  SpecialBiddingMove,
+  Player,
   Suit,
-  Trick
+  Trick,
+  LegalPlayingMove
 } from "../../game-engine/gameStateTypes";
 import {gameStateInit} from "../../game-engine/gameStateInit";
-import {applyEvent, GameStateModifier} from "../../game-engine/gameStateModifiers";
-import {inboundGameEventParser, outboundGameEventConverter} from "../../websocket/events/game";
+import {GameStateModifier, applyMessage} from "../../game-engine/gameStateModifiers";
+import {inboundGameMessageParser, outboundGameEventConverter} from "../../websocket/messages/game";
 import {RouteComponentProps, withRouter} from "react-router";
 import {playerIndexFromPosition} from "../../game-engine/playerPositionning";
 
-const CLEAN_TRICK_TIMOUT_MS = 2000;
+const UPDATE_TRICK_TIMOUT_MS = 2000;
 
-const isFull = (trick: Trick): boolean => !!trick.top && !!trick.bottom && !!trick.left && !!trick.right;
+const hasChanged = (prevTrick: Trick, currTrick: Trick): boolean => {
+  let i = 0;
+  for (const username in prevTrick.cards) {
+    if (Object.prototype.hasOwnProperty.call(prevTrick.cards, username)) {
```

Comment:
Wow I've never seen this, I trust you if it works!

Reply:
I first saw it on stackoverflow and was reluctant to use it because it seems hacky, but my vscode extension actually had a snippet for it so then I was convinced.